File: demos/cpp/data_structures_demo.cpp
#include <iostream>
#include <map>
#include <memory>
#include <queue>
#include <string>
#include <vector>

struct Node {
    std::string value;
    std::vector<std::unique_ptr<Node>> children;
};

int main() {
    std::vector<std::string> steps{"parse", "validate", "render"};
    std::map<std::string, int> scores{{"Ada", 95}, {"Lin", 91}};
    std::priority_queue<int, std::vector<int>, std::greater<>> tasks;
    tasks.push(2);
    tasks.push(1);
    auto tree = std::make_unique<Node>(Node{"root", {}});
    tree->children.push_back(std::make_unique<Node>(Node{"child", {}}));

    std::cout << "vector=" << steps.size() << ':' << steps.front() << '\n';
    std::cout << "map=" << scores.begin()->first << ':' << scores.begin()->second << '\n';
    std::cout << "heap-min=" << tasks.top() << '\n';
    std::cout << "tree-children=" << tree->children.size() << '\n';
}

File: demos/kotlin/data_structures_demo.kt
import java.util.PriorityQueue

data class Task(val name: String, val priority: Int)
sealed interface Tree<out T> {
    data class Leaf<T>(val value: T) : Tree<T>
    data class Branch<T>(val value: T, val children: List<Tree<T>>) : Tree<T>
}

fun main() {
    val values: List<String> = listOf("parse", "validate", "render")
    val mutable = ArrayDeque(values)
    val scores = linkedMapOf("Ada" to 95, "Lin" to 91)
    val tasks = PriorityQueue<Task>(compareBy(Task::priority))
    tasks += Task("docs", 2)
    tasks += Task("tests", 1)
    val tree: Tree<String> = Tree.Branch("root", listOf(Tree.Leaf("child")))

    println("list=${values.joinToString(",")}")
    println("queue-first=${mutable.removeFirst()}")
    println("scores=${scores.entries.joinToString(",") { "${it.key}:${it.value}" }}")
    println("next=${tasks.remove().name}")
    val childCount = (tree as Tree.Branch<String>).children.size
    println("tree=branch:$childCount")
}
